File: fetch-cli-error-handling.c
#include <stdio.h>
// Given a CLI argument which is a filename, print it's contents

int fetch(char* fname) {
    // open fname
    FILE *fp = fopen(fname, "r");
    if (!fp) {
        printf("fetch: %s: 404 Not Found\n", fname);
        return 1;
    }

    // read from fname and print
    char con[100];                          // store content in a char array
    while (fgets(con, 100, fp) != NULL)     // read from file into array
        printf("%s", con);                  // print content

    // close fname
    fclose(fp);
    return 0;
}

int main(int argc, char *argv[]) {
    if ( argc < 2 ) {
        printf("\n  USAGE: fetch FILE\n\n");
        return 1;
    }

    return fetch(argv[1]);
}

/*
# This: fetch-cli-error-handling.c
# Prev: fetch-function-error-handling.c
# Next: learn-about-sockets.md

# Build & run:
gcc fetch-function.c -o fetch && ./fetch
*/

File: java/FetchCliErrorHandling.java
import java.io.*;
// FileInputStream
// FileNotFoundException
// IOException

public class FetchCliErrorHandling {

    private static void fetch(String fname) {
        try{
            FileInputStream fis = new FileInputStream(fname);
            int i;
            byte buf[] = new byte[100];
            while( (i = fis.read(buf))>0 ){
                String str = new String(buf,0,i);
                System.out.print(str);
            }
            fis.close();
        }catch( FileNotFoundException e ){
            System.out.println("404 Not Found: "+fname);
        }catch( IOException e ){
            System.out.println("500 Server Error: Read failed");
        }
    }

    public static void main(String[] args) {
        if( args.length>0 ){
            fetch(args[0]);
        } else {
            System.out.println(
                "\n  USAGE: java FetchCliErrorHandling FILE\n");
        }
    }
}

/*
# This: FetchCliErrorHandling.java
# Prev: FetchFunctionErrorHandling.java
# Next: ServerInitial.java

# Build & run:
JNAME=FetchCliErrorHandling; javac ${JNAME}.java && java ${JNAME}
*/

